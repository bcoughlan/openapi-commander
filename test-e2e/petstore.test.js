const util = require('node:util')
const exec = util.promisify(require('node:child_process').exec)
const fs = require('fs/promises')
const express = require('express');

//TODO generate unique tmp dir.
//TODO Use any free port for the server

const baseUrl = `http://localhost:3123`

describe('Petstore', () => {

  let server
  beforeAll(async () => {
    const app = express()
    app.use(express.json())
    app.get('/pet/1', (req, res) => {
      res.json({ id: 1, name: 'doggie' })
    })
    app.get('/pet/2', (req, res) => {
      res.json({ auth: req.headers['authorization'] })
    })
    app.get('/customserver/pet/1', (req, res) => {
      res.json({ msg: 'custom server worked' })
    })
    app.put('/pet', (req, res) => {
      res.json(req.body)
    })
    server = app.listen(3123)

    await exec('mkdir -p tmp && node bin/run.js generate test-resources/petstore.yaml tmp/petstore.js')
  })

  afterAll(async () => {
    server.close()
    await exec('rm -rf /tmp/petstore.js')
  })

  it('lists categories', async () => {
    expect(await run('')).toMatchSnapshot()
  })

  it('lists commands', async () => {
    expect(await run('pet')).toMatchSnapshot()
  })

  it('can do a GET', async () => {
    expect(await run('pet getPetById 1')).toMatchSnapshot()
  })

  it('debug prints', async () => {
    expect(await run('-d pet getPetById 10')).toMatchSnapshot()
  })

  it('prints examples', async () => {
    expect(await run('pet updatePet examples')).toMatchSnapshot()
  })

  it('prints --help on a PUT', async () => {
    expect(await run('pet updatePet --help')).toMatchSnapshot()
  })

  it('can do a PUT', async () => {
    await fs.writeFile('/tmp/pet.json', '{"id":10, "name":"test", "status":"available"}')
    expect(await run('pet updatePet /tmp/pet.json')).toMatchSnapshot()
  })

  describe('auth header', () => {
    it('can set with option', async () => {
      const { stdout } = await run('pet getPetById 2 -a "test auth"')
      expect(stdout).toContain('"test auth"')
    })

    it('can set with env var', async () => {
      const { stdout } = await run('pet getPetById 2', { env: { 'PETSTORE_AUTH': 'test env auth' } })
      expect(stdout).toContain('"test env auth"')
    })

    it('gives precedence to option over env var', async () => {
      const { stdout } = await run('pet getPetById 2 -a "test auth"', { env: { 'PETSTORE_AUTH': 'test env auth' } })
      expect(stdout).toContain('"test auth"')
    })
  })

  describe('server', () => {
    it('can set with option', async () => {
      const { stdout } = await run('pet getPetById 1', { server: `${baseUrl}/customserver` })
      expect(stdout).toContain('"custom server worked"')
    })

    it('can set with env var', async () => {
      const { stdout } = await run('pet getPetById 1', { server: null, env: { 'PETSTORE_SERVER': `${baseUrl}/customserver` } })
      expect(stdout).toContain('"custom server worked"')
    })

    it('can set with env var', async () => {
      const { stdout } = await run('pet getPetById 1', { server: `${baseUrl}/customserver`, env: { 'PETSTORE_SERVER': `${baseUrl}/badpath` } })
      expect(stdout).toContain('"custom server worked"')
    })
  })

  //TODO server

  async function run(cmd, { env = {}, server = baseUrl } = {}) {
    try {
      let { stdout, stderr } = await exec(`node tmp/petstore.js ${!server ? '' : `-s ${server}`} ${cmd} 2>&1`, { env: { ...process.env, ...env } })
      return { stdout, stderr }
    } catch (e) {
      return { stdout: e.stdout, stderr: e.stderr, code: e.code }
    }
  }
})
