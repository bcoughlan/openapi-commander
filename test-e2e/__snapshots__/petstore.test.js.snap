// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Petstore can do a GET 1`] = `
Object {
  "stderr": "",
  "stdout": "Status: 200
{
  \\"id\\": 1,
  \\"name\\": \\"doggie\\"
}
",
}
`;

exports[`Petstore can do a PUT 1`] = `
Object {
  "stderr": "",
  "stdout": "Status: 200
{
  \\"id\\": 10,
  \\"name\\": \\"test\\",
  \\"status\\": \\"available\\"
}
",
}
`;

exports[`Petstore debug prints 1`] = `
Object {
  "stderr": "",
  "stdout": "GET http://localhost:3123/pet/10
Accept: */*
",
}
`;

exports[`Petstore lists categories 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Usage: petstore [options] [command]

Options:
  -p, --print <mode>     Print the HTTP request instead of sending it.
                         (choices: \\"curl\\", \\"plain\\")
  -v, --verbose          Includes the response headers in the output
  -s, --server <server>  Base URL to use for requests
  -a, --auth <auth>      Authorization header to send
  -h, --help             display help for command

Commands:
  pet                    Everything about your Pets
  store
  user                   Operations about user
  help [command]         display help for command
",
}
`;

exports[`Petstore lists commands 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Usage: petstore pet [options] [command]

Everything about your Pets

Options:
  -h, --help                           display help for command

Commands:
  updatePet [options] <body>           Update an existing pet
  addPet [options] <body>              Add a new pet to the store
  findPetsByStatus [options]           Finds Pets by status
  findPetsByTags [options]             *DEPRECATED* Finds Pets by tags
  getPetById <petId>                   Find pet by ID
  updatePetWithForm [options] <petId>  Updates a pet in the store with form
                                       data
  deletePet [options] <petId>          Deletes a pet
  uploadFile [options] <petId>         uploads an image
  help [command]                       display help for command
",
}
`;

exports[`Petstore prints --help on a PUT 1`] = `
Object {
  "stderr": "",
  "stdout": "Usage: petstore pet updatePet [options] [command] <body>

Update an existing pet by Id

Arguments:
  body                    Path to a file containing the request body.

Options:
  --body-type <bodyType>  Content type of the request body. (choices:
                          \\"application/json\\", \\"application/xml\\",
                          \\"application/x-www-form-urlencoded\\", default:
                          \\"application/json\\")
  -h, --help              display help for command

Commands:
  examples
",
}
`;

exports[`Petstore prints a curl GET request 1`] = `
Object {
  "stderr": "",
  "stdout": "curl -X GET -H 'Accept: */*' 'http://localhost:3123/pet/10'
",
}
`;

exports[`Petstore prints a curl POST request 1`] = `
Object {
  "stderr": "",
  "stdout": "curl -X POST -H 'Authorization: Bearer mytoken' -H 'Content-Type: application/json' -H 'Accept: */*' 'http://localhost:3123/pet' -d '{\\"id\\":1,\\"name\\":\\"doggie\\"}'
",
}
`;

exports[`Petstore prints examples 1`] = `
Object {
  "stderr": "",
  "stdout": "Example for application/json:
{
  [36m\\"id\\"[39m: [32m10[39m,
  [36m\\"name\\"[39m: [31m\\"doggie\\"[39m,
  [36m\\"category\\"[39m: {
    [36m\\"id\\"[39m: [32m1[39m,
    [36m\\"name\\"[39m: [31m\\"Dogs\\"[39m
  },
  [36m\\"photoUrls\\"[39m: [
    [31m\\"string\\"[39m
  ],
  [36m\\"tags\\"[39m: [
    {
      [36m\\"id\\"[39m: [32m0[39m,
      [36m\\"name\\"[39m: [31m\\"string\\"[39m
    }
  ],
  [36m\\"status\\"[39m: [31m\\"available\\"[39m
}
Example for application/xml:
{
  id: 10,
  name: 'doggie',
  category: { id: 1, name: 'Dogs' },
  photoUrls: [ 'string' ],
  tags: [ { id: 0, name: 'string' } ],
  status: 'available'
}
Example for application/x-www-form-urlencoded:
{
  id: 10,
  name: 'doggie',
  category: { id: 1, name: 'Dogs' },
  photoUrls: [ 'string' ],
  tags: [ { id: 0, name: 'string' } ],
  status: 'available'
}
",
}
`;
